agent/grpc-internal/listener.go

```diff
@@ -0,0 +1,60 @@
+package internal
+
+import (
+	"fmt"
+	"net"
+)
+
+// PipeListener implements the net.Listener interface and allows you to manually
+// pass connections to it. This is useful when you need to accept connections
+// and do something with them yourself first (e.g. handling our multiplexing
+// scheme) before giving them to the gRPC server.
+type PipeListener struct {
+	addr  net.Addr
+	conns chan net.Conn
+	done  chan struct{}
+}
+
+var _ net.Listener = (*PipeListener)(nil)
+
+// NewPipeListener creates a PipeListener with the given address.
+func NewPipeListener(addr net.Addr) *PipeListener {
+	return &PipeListener{
+		addr:  addr,
+		conns: make(chan net.Conn),
+		done:  make(chan struct{}),
+	}
+
+}
+
+// Handle makes the given connection available to Accept.
+func (l *PipeListener) Handle(conn net.Conn) {
+	select {
+	case l.conns <- conn:
+	case <-l.done:
+	}
+}
```

Comment:
As this could (at least in theory) block indefinitely if nothing is calling `Accept` on the listener, I think it would be advantageous to add a `context.Context` in the signature and to allow for this to return an error indicating the conn was not successfully processed.

I suspect this would trickle down to other RPC multiplex handling code passing in a new context with some timeout and handling the error by logging the timeout.